centreon/www/include/configuration/configObject/host/DB-Func.php

```diff
@@ -3118,23 +3205,108 @@ function deleteHostFromVault(
+        }
+    }
+
+    //Unset existing macros on vault if they no more exist while submitting the form
+    foreach(array_keys($hostSecretsFromVault) as $secretKey) {
+        if ($secretKey !== SNMP_COMMUNITY_MACRO_NAME) {
+            $macroName = [];
+            foreach($macros as $macroInfos) {
+                $macroName[] = $macroInfos['macroName'];
+                if (array_key_exists('originalName', $macroInfos) && $secretKey === $macroInfos['originalName']) {
+                    $hostSecretsFromVault[$macroInfos['macroName']] = $hostSecretsFromVault[$secretKey];
+                }
+            }
+            if (! in_array($secretKey, $macroName)) {
+                unset($hostSecretsFromVault[$secretKey]);
+            }
+        }
+    }
+
+    //Unset existing SNMP Community if it no more exists while submitting the form
     if (
-        array_key_exists(':host_snmp_community', $bindParams)
-        && ! preg_match('/^secret::\d+::/', $bindParams[':host_snmp_community'][\PDO::PARAM_STR])
-        && $bindParams[':host_snmp_community'][\PDO::PARAM_STR] !== null
+        array_key_exists(SNMP_COMMUNITY_MACRO_NAME, $hostSecretsFromVault)
+        && $hostSNMPCommunity === null
     ) {
-        return [
-            '_HOSTSNMPCOMMUNITY' => $bindParams[':host_snmp_community'][\PDO::PARAM_STR]
-        ];
+        unset($hostSecretsFromVault[SNMP_COMMUNITY_MACRO_NAME]);
+    }
+
+    //Add SNMP Community if a new value has been set
+    if ($hostSNMPCommunity !== null && ! preg_match(VAULT_PATH_REGEX, $hostSNMPCommunity)) {
+        $formPasswordValues[SNMP_COMMUNITY_MACRO_NAME] = $hostSNMPCommunity;
+    }
+
+    $result = array_diff($formPasswordValues, $hostSecretsFromVault);
+
+    return array_merge($hostSecretsFromVault, $result);
+}
+
+/**
+ * Add new macros and SNMP Community to the write in vault payload
+ *
+ * @param string|null $hostSNMPCommunity
+ * @param array<int,array<string,string>> $macros
+ * @param array<string,string> $hostSecretsFromVault
+ * @return array<string,string>
+ */
+function prepareUpdateMCPayload(?string $hostSNMPCommunity, array $macros, array $hostSecretsFromVault)
+{
+    foreach($macros as $macroInfos) {
+        if (
+            $macroInfos['macroPassword'] === '1'
+            && ! preg_match(VAULT_PATH_REGEX, $macroInfos['macroValue'])
+        ) {
+            $hostSecretsFromVault[$macroInfos['macroName']] = $macroInfos['macroValue'];
+        }
+    }
+
+    //Add SNMP Community if a new value has been set
+    if ($hostSNMPCommunity !== null && ! preg_match(VAULT_PATH_REGEX, $hostSNMPCommunity)) {
+        $hostSecretsFromVault[SNMP_COMMUNITY_MACRO_NAME] = $hostSNMPCommunity;
+    }
+
+    return $hostSecretsFromVault;
+}
+
+/**
+ * Store all the ids of password macros that have been updated
+ *
+ * @param array<int,array<string,string> $macros
+ * @return int[]
+ */
+function getIdOfUpdatedPasswordMacros(array $macros): array
```

Comment:
```suggestion
function getIdOfUpdatedPasswordMacros(array $macros): array
```
use this function everywhere its needed